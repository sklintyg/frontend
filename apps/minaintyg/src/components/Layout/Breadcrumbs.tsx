import { IDSBreadcrumbs, IDSCrumb } from '@frontend/ids-react-ts'
import { Link, Params, useMatches } from 'react-router-dom'

type Match = ReturnType<typeof useMatches>[number]
type MatchWithCrumb = Match & {
  handle: { crumb: (data: Params<string>) => string }
}

function hasCrumb(match: MatchWithCrumb | Match): match is MatchWithCrumb {
  return (match as MatchWithCrumb).handle !== undefined && typeof (match as MatchWithCrumb).handle.crumb === 'function'
}

function getMatchAt(index: number, matches: MatchWithCrumb[]): { pathname: string; crumb: string } | null {
  const match = matches.at(index)
  return match ? { pathname: match.pathname, crumb: match.handle.crumb(match.params) } : null
}

export function Breadcrumbs() {
  const matches = useMatches().filter(hasCrumb)
  const prevMatch = getMatchAt(-2, matches)
  const currentMatch = getMatchAt(-1, matches)

  return (
    <div className="mb-5">
      <IDSBreadcrumbs current={currentMatch?.crumb ?? 'Start'} srlabel="Du 채r h채r" lead="Du 채r h채r:">
        {matches.slice(0, -1).map(({ handle, params, pathname }) => (
          <IDSCrumb key={pathname}>
            <Link to={pathname}>{handle.crumb(params)}</Link>
          </IDSCrumb>
        ))}
        {prevMatch && (
          <IDSCrumb key="mobile" mobile>
            <Link to={prevMatch.pathname}>{prevMatch.crumb}</Link>
          </IDSCrumb>
        )}
      </IDSBreadcrumbs>
    </div>
  )
}
