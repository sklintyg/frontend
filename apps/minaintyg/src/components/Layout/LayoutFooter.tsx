import { IDSFooter, IDSIconChevronBold, IDSIconExternal, IDSLink, IDSMobileMenu, IDSMobileMenuItem } from '@frontend/ids-react-ts'
import navigation from '../../data/1177-navbar-services.json'
import { resolveNavigationUrl } from '../../utils/resolveNavigationUrl'
import { LayoutFooterLink } from './LayoutFooterLink'

const footerLinks = [
  {
    col: '2',
    links: [
      ['Support', 'https://www.1177.se/e-tjanster-support', '_blank'],
      [
        'Tillgänglighet',
        'https://www.1177.se/om-1177/1177.se/tillganglighet-pa-1177/tillganglighetsredogorelse-for-e-tjanster-pa-1177.se/',
        '_blank',
      ],
    ],
  },
  {
    col: '3',
    links: [
      ['Läs mer om e-tjänster', 'https://www.1177.se/e-tjanster', '_blank'],
      ['Webbkarta', 'https://e-tjanster.1177.se/mvk/sitemap.xhtml', '_self'],
    ],
  },
]

export function LayoutFooter() {
  const startLinkItem = navigation.menu.items.find(({ name }) => name === 'Start')
  return (
    <IDSFooter type="1177" headline="1177" subheadline="– tryggt om din hälsa och vård" className="print:hidden">
      <p>1177 är en tjänst från Sveriges regioner. Vi finns alltid med dig när du vill må bättre.</p>

      {footerLinks.map(({ col, links }) =>
        links.map(([name, url, target]) => (
          <IDSLink key={`${col}-${name}`} slot={`link-col-${col}`}>
            <IDSIconChevronBold />
            <a href={url} target={target}>
              {name}
            </a>
            {target === '_blank' && <IDSIconExternal slot="append-icon" />}
          </IDSLink>
        ))
      )}

      <span slot="sub-footer-right" className="relative">
        <LayoutFooterLink href="https://www.1177.se/om-1177/1177.se/hantering-av-personuppgifter-pa-1177.se/" target="_blank">
          Behandling av personuppgifter
        </LayoutFooterLink>
      </span>
      <span slot="sub-footer-right" className="relative">
        <LayoutFooterLink href="https://www.1177.se/om-1177/1177.se/hantering-av-kakor-cookies-pa-1177.se/" target="_blank">
          Hantering av kakor
        </LayoutFooterLink>
      </span>

      {startLinkItem &&
        ['left', 'mobile'].map((slot) => (
          <span key={slot} slot={`sub-footer-${slot}`} className="relative">
            1177 drivs av{' '}
            <LayoutFooterLink href={resolveNavigationUrl(startLinkItem.url)} target="_blank">
              Inera AB
            </LayoutFooterLink>{' '}
            på uppdrag av Sveriges regioner.
          </span>
        ))}

      <IDSMobileMenu>
        {footerLinks.map(({ col, links }) =>
          links.map(([name, url, target]) => (
            <IDSMobileMenuItem key={`${col}-${name}`}>
              <a href={url} target={target} rel="noreferrer" className="relative">
                {name}
                {target === '_blank' && (
                  <IDSIconExternal
                    className="ids-ml-2 relative -bottom-0.5"
                    inline
                    color="var(--mobile-menu-item-icon_color)"
                    height="100%"
                    width="17"
                  />
                )}
              </a>
            </IDSMobileMenuItem>
          ))
        )}
      </IDSMobileMenu>

      <span slot="sub-footer-links">
        <LayoutFooterLink href="https://www.1177.se/om-1177/1177.se/hantering-av-personuppgifter-pa-1177.se/" target="_blank">
          Behandling av personuppgifter
        </LayoutFooterLink>
      </span>
      <span slot="sub-footer-links">
        <LayoutFooterLink href="https://www.1177.se/om-1177/1177.se/hantering-av-kakor-cookies-pa-1177.se/" target="_blank">
          Hantering av kakor
        </LayoutFooterLink>
      </span>
    </IDSFooter>
  )
}
