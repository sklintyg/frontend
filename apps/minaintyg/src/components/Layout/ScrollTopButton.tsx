import { classNames } from '@frontend/components'
import { IDSButton, IDSIconArrow } from '@frontend/ids-react-ts'
import { getTabbableNodes } from 'focus-lock'
import { useEffect, useState } from 'react'

export function ScrollTopButton() {
  const [show, setShow] = useState(false)
  const [isScrolling, updateIsScrolling] = useState(false)

  useEffect(() => {
    let previousScrollPos = 0
    const handleScroll = () => {
      const scrollPos = document.body.getBoundingClientRect().top
      const isScrollInterupted = previousScrollPos > 0 && scrollPos > previousScrollPos

      setShow(scrollPos < -180)
      if (isScrolling && (scrollPos === 0 || isScrollInterupted)) {
        updateIsScrolling(false)
        const mainElement = document.querySelector('main')
        if (!isScrollInterupted && mainElement instanceof HTMLElement) {
          getTabbableNodes([mainElement], new Map())[0].node.focus()
        }
        previousScrollPos = 0
      }
      previousScrollPos = scrollPos
    }

    window.addEventListener('scroll', handleScroll, true)
    return () => {
      window.removeEventListener('scroll', handleScroll, true)
    }
  }, [isScrolling])

  return (
    <div
      className={classNames(
        'sticky bottom-0 h-16 opacity-0 transition-opacity ease-in-out pointer-events-none',
        show && 'opacity-100 pointer-events-auto'
      )}
    >
      <IDSButton
        fab
        role="button"
        className="ids-hide ids-show-1177 absolute bottom-5 right-5 select-none"
        aria-label="Till toppen av sidan"
        onClick={() => {
          updateIsScrolling(true)
          window.scrollTo({
            top: 0,
            left: 0,
            behavior: 'smooth',
          })
        }}
      >
        <IDSIconArrow color="var(--color-main)" width="24" height="24" rotate="-90" />
      </IDSButton>
    </div>
  )
}
