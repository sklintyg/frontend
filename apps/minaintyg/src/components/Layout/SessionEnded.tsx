import { IDSIconChevron, IDSIconExternal, IDSLink } from '@frontend/ids-react-ts'
import { Link } from 'react-router-dom'
import { PageHero } from '../PageHero/PageHero'
import { PageHeroActions } from '../PageHero/PageHeroActions'

export function SessionEnded() {
  return (
    <PageHero heading="Du 채r utloggad" type="success">
      <p className="ids-preamble">
        Du har antingen valt att logga ut eller blivit utloggad p책 grund av inaktivitet. Du kan nu st채nga fliken eller logga in p책 nytt.
      </p>
      <PageHeroActions>
        <IDSLink>
          <IDSIconChevron />
          <Link to={import.meta.env.VITE_LOGIN_URL}>Till inloggning</Link>
        </IDSLink>
        <IDSLink>
          <IDSIconChevron />
          <Link to="https://www.1177.se">Till 1177</Link>
          <IDSIconExternal slot="append-icon" />
        </IDSLink>
      </PageHeroActions>
    </PageHero>
  )
}
