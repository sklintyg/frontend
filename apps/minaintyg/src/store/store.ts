import { configureStore } from '@reduxjs/toolkit'
import { api } from './api'
import { errorMiddleware } from './middleware/error.middleware'
import { reducer } from './reducer'
import { idpMiddleware } from './slice/idpMiddleware'

export const store = configureStore({
  reducer,
  middleware: (getDefaultMiddleware) => getDefaultMiddleware().concat([api.middleware, idpMiddleware, errorMiddleware]),
})

export type AppDispatch = typeof store.dispatch
