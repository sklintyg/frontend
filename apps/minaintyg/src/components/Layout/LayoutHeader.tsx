/* eslint-disable jsx-a11y/anchor-is-valid */
import { HeaderNavItem, MobileMenuItem } from '@frontend/components'
import { IDSHeader, IDSHeaderAvatar, IDSHeaderMobileItem, IDSHeaderMobileMenu, IDSHeaderNav } from '@frontend/ids-react-ts'
import { Link } from 'react-router-dom'
import { useLogout } from '../../hooks/useLogout'
import { useGetUserQuery } from '../../store/api'

const links = [
  { title: 'Start', path: '/' },
  { title: 'Inkorg', path: '/inkorg' },
  { title: 'Bokade tider', path: '/bokade-tider' },
  { title: 'Journalen', path: '/journalen' },
  { title: 'Egen provhantering', path: '/egen-provhantering' },
  { title: 'Stöd och behandling', path: '/stod-och-behandling' },
  { title: 'Intyg', path: '/intyg' },
  { title: 'Övriga tjänster', path: '/ovriga-tjanster' },
]

export function LayoutHeader() {
  const logout = useLogout()
  const { data: user } = useGetUserQuery()

  return (
    <IDSHeader type="1177" hideregionpicker className="z-40 bg-white print:hidden">
      {user && (
        <>
          <IDSHeaderAvatar username={user?.personName ?? ''}>
            <button type="button" slot="avatar-left">
              Inställningar
            </button>
            <button type="button" slot="avatar-right" onClick={logout}>
              Logga ut
            </button>
          </IDSHeaderAvatar>

          <IDSHeaderNav>
            {links.map(({ title, path }) => (
              <HeaderNavItem key={path} to={path} title={title} />
            ))}

            <IDSHeaderMobileItem>
              <Link to="/installningar">Inställningar</Link>
            </IDSHeaderMobileItem>
            <IDSHeaderMobileItem>
              <a href="#" onClick={logout}>
                Logga ut
              </a>
            </IDSHeaderMobileItem>

            <IDSHeaderMobileMenu type="1177">
              Meny
              {links.map(({ title, path }) => (
                <MobileMenuItem key={path} to={path} title={title} />
              ))}
            </IDSHeaderMobileMenu>
          </IDSHeaderNav>
        </>
      )}
    </IDSHeader>
  )
}
