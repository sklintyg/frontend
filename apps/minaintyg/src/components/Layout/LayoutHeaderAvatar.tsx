import { HeaderAvatarLink } from '@frontend/components'
import { IDSHeader1177Avatar } from '@inera/ids-react'
import { useGetUserQuery } from '../../store/hooks'
import { useLinks } from '../../hooks/useLinks'

export function LayoutHeaderAvatar() {
  const { data: user } = useGetUserQuery()
  const settingLink = useLinks().find((link) => link.id === '99')

  if (!user) {
    return null
  }

  return (
    <IDSHeader1177Avatar
      username={user.personName}
      links={
        <>
          {settingLink && (
            <HeaderAvatarLink icon="settings" to={settingLink.url}>
              {settingLink.name}
            </HeaderAvatarLink>
          )}
          <HeaderAvatarLink icon="user" to="/logga-ut">
            Logga ut
          </HeaderAvatarLink>
        </>
      }
    />
  )
}
