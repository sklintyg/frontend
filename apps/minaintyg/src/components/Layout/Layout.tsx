import { GlobalAlert, PriorityEnum } from '@frontend/components'
import { LayoutFooter, LayoutHeader, LayoutHeaderNavigation } from '@frontend/components/1177'
import { ReactNode, useRef } from 'react'
import { useDocumentTitle } from '../../hooks/useDocumentTitle'
import { BannerPriority } from '../../schema/informationSchema'
import { useGetInfoQuery } from '../../store/api'
import { useAppSelector, useGetUserQuery } from '../../store/hooks'
import { ErrorPageHero } from '../error/ErrorPageHero'
import { LayoutHeaderAvatar } from './LayoutHeaderAvatar'
import { ScrollTopButton } from './ScrollTopButton'

export function Layout({ children }: { children: ReactNode }) {
  const { hasSessionEnded, reason, errorId } = useAppSelector((state) => state.sessionSlice)
  const ref = useRef<HTMLDivElement>(null)
  useDocumentTitle(ref)
  const { data: user } = useGetUserQuery()
  const hasSession = useAppSelector((state) => state.sessionSlice.hasSession)
  const { data: info } = useGetInfoQuery()
  const getAlertPriority = (priority: BannerPriority) => {
    if (priority === BannerPriority.ERROR) {
      return PriorityEnum.ERROR
    }
    return priority === BannerPriority.OBSERVE ? PriorityEnum.OBSERVE : PriorityEnum.INFO
  }
  return (
    <div id="top" className="flex min-h-screen flex-col">
      <LayoutHeader mode={import.meta.env.MODE}>
        {user && (
          <>
            <LayoutHeaderAvatar />
            <LayoutHeaderNavigation mode={import.meta.env.MODE} activeLink="Intyg" />
          </>
        )}
      </LayoutHeader>
      <main id="content" className="relative flex-1">
        {info &&
          info.banners.length > 0 &&
          info.banners.map((banner) => (
            <GlobalAlert key={banner.type} priority={getAlertPriority(banner.type)}>
              {banner.content}
            </GlobalAlert>
          ))}
        <div ref={ref} className="ids-content m-auto max-w-screen-xl overflow-hidden px-2.5 py-5">
          {hasSessionEnded ? <ErrorPageHero type={reason} id={errorId} /> : children}
        </div>
        <ScrollTopButton />
      </main>
      <LayoutFooter hasSession={hasSession} />
    </div>
  )
}
