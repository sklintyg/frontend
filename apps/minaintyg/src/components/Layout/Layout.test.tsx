import { act, render } from '@testing-library/react'
import { Provider } from 'react-redux'
import { Route, RouterProvider, createMemoryRouter, createRoutesFromChildren } from 'react-router-dom'
import { waitForRequest } from '../../mocks/server'
import { store } from '../../store/store'
import { Layout } from './Layout'

it('Should render as expected', async () => {
  const userRequest = waitForRequest('GET', '/api/user')

  const { baseElement } = render(
    <Provider store={store}>
      <RouterProvider
        router={createMemoryRouter(
          createRoutesFromChildren([
            <Route key="root" path="/" element={<Layout />}>
              <Route index element={<p>FooBar</p>} />
            </Route>,
          ]),
          { initialEntries: ['/'] }
        )}
      />
    </Provider>
  )

  await act(async () => userRequest)

  expect(baseElement).toMatchSnapshot()
})
