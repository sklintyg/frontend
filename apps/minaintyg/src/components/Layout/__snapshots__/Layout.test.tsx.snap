// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Should render as expected 1`] = `
<body>
  <div>
    <div
      class="flex min-h-screen flex-col"
    >
      <ids-header
        class="z-40 bg-white print:hidden"
        type="1177"
      >
        <ids-header-avatar
          type="1177"
        >
          <button
            slot="avatar-left"
            type="button"
          >
            Inställningar
          </button>
          <button
            slot="avatar-right"
            type="button"
          >
            Logga ut
          </button>
        </ids-header-avatar>
        <ids-header-nav
          slot="header-nav"
          type="1177"
        >
          <ids-header-nav-item
            type="1177"
          >
            <a
              href="/"
            >
              Start
            </a>
          </ids-header-nav-item>
          <ids-header-nav-item
            type="1177"
          >
            <a
              href="/inkorg"
            >
              Inkorg
            </a>
          </ids-header-nav-item>
          <ids-header-nav-item
            type="1177"
          >
            <a
              href="/bokade-tider"
            >
              Bokade tider
            </a>
          </ids-header-nav-item>
          <ids-header-nav-item
            type="1177"
          >
            <a
              href="/journalen"
            >
              Journalen
            </a>
          </ids-header-nav-item>
          <ids-header-nav-item
            type="1177"
          >
            <a
              href="/egen-provhantering"
            >
              Egen provhantering
            </a>
          </ids-header-nav-item>
          <ids-header-nav-item
            type="1177"
          >
            <a
              href="/stod-och-behandling"
            >
              Stöd och behandling
            </a>
          </ids-header-nav-item>
          <ids-header-nav-item
            type="1177"
          >
            <a
              href="/intyg"
            >
              Intyg
            </a>
          </ids-header-nav-item>
          <ids-header-nav-item
            type="1177"
          >
            <a
              href="/ovriga-tjanster"
            >
              Övriga tjänster
            </a>
          </ids-header-nav-item>
          <ids-header-mobile-item>
            <a
              href="/installningar"
            >
              Inställningar
            </a>
          </ids-header-mobile-item>
          <ids-header-mobile-item>
            <a
              href="#"
            >
              Logga ut
            </a>
          </ids-header-mobile-item>
          <ids-header-mobile-menu
            type="1177"
          >
            Meny
            <ids-mobile-menu-item
              active=""
              slot="menu-link"
            >
              <a
                href="/"
                slot="link"
              >
                Start
              </a>
            </ids-mobile-menu-item>
            <ids-mobile-menu-item
              slot="menu-link"
            >
              <a
                href="/inkorg"
                slot="link"
              >
                Inkorg
              </a>
            </ids-mobile-menu-item>
            <ids-mobile-menu-item
              slot="menu-link"
            >
              <a
                href="/bokade-tider"
                slot="link"
              >
                Bokade tider
              </a>
            </ids-mobile-menu-item>
            <ids-mobile-menu-item
              slot="menu-link"
            >
              <a
                href="/journalen"
                slot="link"
              >
                Journalen
              </a>
            </ids-mobile-menu-item>
            <ids-mobile-menu-item
              slot="menu-link"
            >
              <a
                href="/egen-provhantering"
                slot="link"
              >
                Egen provhantering
              </a>
            </ids-mobile-menu-item>
            <ids-mobile-menu-item
              slot="menu-link"
            >
              <a
                href="/stod-och-behandling"
                slot="link"
              >
                Stöd och behandling
              </a>
            </ids-mobile-menu-item>
            <ids-mobile-menu-item
              slot="menu-link"
            >
              <a
                href="/intyg"
                slot="link"
              >
                Intyg
              </a>
            </ids-mobile-menu-item>
            <ids-mobile-menu-item
              slot="menu-link"
            >
              <a
                href="/ovriga-tjanster"
                slot="link"
              >
                Övriga tjänster
              </a>
            </ids-mobile-menu-item>
          </ids-header-mobile-menu>
        </ids-header-nav>
      </ids-header>
      <main
        class="relative flex-1"
      >
        <div
          class="ids-content m-auto max-w-7xl overflow-hidden p-5"
        >
          <div
            class="mb-5"
          >
            <ids-breadcrumbs />
          </div>
          <p>
            FooBar
          </p>
        </div>
        <div
          class="sticky bottom-0 h-16 opacity-0 transition-opacity ease-in-out"
        >
          <ids-button
            aria-disabled="false"
            class="ids-hide ids-show-1177 absolute bottom-5 right-5 select-none"
            fab=""
            role="button"
            size="m"
            tabindex="0"
            type=""
          >
            <ids-icon-arrow
              class="ids-icon"
              size="s"
            />
          </ids-button>
        </div>
      </main>
      <ids-footer
        class="print:hidden ids-type-1177"
        cols="3"
        linkcol1="2"
        linkcol2="2"
        linkcol3="2"
        type="1177"
      />
    </div>
  </div>
</body>
`;

exports[`Should render without header when user is not logged in 1`] = `
<body>
  <div>
    <div
      class="flex min-h-screen flex-col"
    >
      <ids-header
        class="z-40 bg-white print:hidden"
        type="1177"
      />
      <main
        class="relative flex-1"
      >
        <div
          class="ids-content m-auto max-w-7xl overflow-hidden p-5"
        >
          <div
            class="mb-5"
          >
            <ids-breadcrumbs />
          </div>
          <p>
            FooBar
          </p>
        </div>
        <div
          class="sticky bottom-0 h-16 opacity-0 transition-opacity ease-in-out"
        >
          <ids-button
            aria-disabled="false"
            class="ids-hide ids-show-1177 absolute bottom-5 right-5 select-none"
            fab=""
            role="button"
            size="m"
            tabindex="0"
            type=""
          >
            <ids-icon-arrow
              class="ids-icon"
              size="s"
            />
          </ids-button>
        </div>
      </main>
      <ids-footer
        class="print:hidden ids-type-1177"
        cols="3"
        linkcol1="2"
        linkcol2="2"
        linkcol3="2"
        type="1177"
      />
    </div>
  </div>
</body>
`;
